Step2 : 키친포스 요구사항 정리

hanseu9839 opened the pull request:
## 작업 내용
- kitchenops 패키지의 내용을 보고 README.md 파일에 요구사항을 정리하였습니다.
- Step1 미션 때 작성해주신 코드 리뷰 사항을 반영하였습니다.

---

## 고민 사항
domain - service - controller 순으로 참고하여 요구사항을 작성했습니다.
모든 이해 관계자 시점에서 작성을 하였는데, 가능하면 어떤 시점이나 시야로 작성해야하는지 궁금합니다.

hanseu9839 committed:
refactor: 리뷰 반영 코드 리팩토링

hanseu9839 committed:
refactor : testCode assertThat -> assertAll 변경

hanseu9839 committed:
docs : Readme.md 요구 사항 작성

hanseu9839 committed:
test: 실패 테스트 코드 작성

hanseu9839 committed:
feat: 테스트 성공 코드 작성 및 리팩토링 구현 완료

hanseu9839 committed:
refactor: Numbers -> NumberGroups 클래스 명 변경

hanseu9839 committed:
refactor: Number 클래스 validate 추가 및 테스트 코드 추가

hanseu9839 committed:
refactor: Pattern 객체 상수화

hanseu9839 committed:
refactor: Number 클래스 상수 , 필드 값 한 줄 구분 추가

hanseu9839 committed:
Merge branch 'hanseu9839' of https://github.com/hanseu9839/ddd-legacy into hanseu9839

hanseu9839 committed:
refactor: 리뷰 반영 코드 리팩토링
- NumberGroups에서 문자열 파싱 로직을 분리하여 NumberGroupsParser로 이동
- NumberGroups는 숫자 리스트와 도메인 로직(합계 등)에만 집중하도록
  구조변경

hanseu9839 committed:
Docs: 키친포스 요구사항 추가

hanseu9839 committed:
git merge

hanseu9839 committed:
refactor: 유효성 검사 메서드 리팩토링
- 반환값이 없는 보이드 메서드로 전환하여 단순 검증만 수행하도록 수정

liquidjoo commented on README.md:
```diff
@@ -28,6 +31,92 @@ docker compose -p kitchenpos up -d
+- 식당에서 음식 주문을 받는 서비스를 구현한다.
+
+### 메뉴(Menu) : 현재 판매될 수 있는 상품의 묶음
+- [ ] 메뉴는 생성/조회가 가능하다.
```
메뉴 생성과 조회를 나누어보면 어떨까요?
각 책임을 고려해서 요구사항을 잘 정리해보면 좋을 것 같아요 :)

hanseu9839 replied:
넵 생성과 조회를 분리하는 것이 더 명확한 요구사항 일 것 같네요..!

liquidjoo commented on README.md:
```diff
@@ -28,6 +31,92 @@ docker compose -p kitchenpos up -d
+- [ ] **메뉴의 화면 표시 여부**는 변경할 수 있으나, **메뉴 가격이 상품 가격 합보다 큰 경우**에는 표시 상태로 바꿀 수 없다.
+
+### 메뉴 그룹 (MenuGroup) : 판매되는 메뉴의 상위 분류 값
+- [ ] 메뉴 그룹은 **생성/조회**가 가능하다.
```
위와 마찬가지로 생성 조회를 따로 정리해보면 좋을 것 같아요

hanseu9839 replied:
넵..!

liquidjoo commented on README.md:
```diff
@@ -28,6 +31,92 @@ docker compose -p kitchenpos up -d
+- [ ] 이름(name)은 **필수 입력 항목**이다.
+
+### 상품 (Product) : 판매되는 상품
+- [ ] **상품 조회**가 가능하다.
```
가능하다 라고 표현도 좋지만 상품 조회를 한다 와 같이 확실히 나타내보는 것도 좋을 것 같네요

liquidjoo replied:
조금 더 나아가 도메인에 어떠한 필드 값들이 존재하는지도 표현해보아요~

hanseu9839 replied:
조금 더 행위적으로 정리해보겠습니다..!

liquidjoo commented on README.md:
```diff
@@ -28,6 +31,92 @@ docker compose -p kitchenpos up -d
+- [ ] 주문에는 총 **6가지 상태**가 존재한다.
+    - 준비(WAITTING), 수락(ACCEPTED), 서빙됨(SERVED), 배달중(DELIVERING), 배달완료(DELIVERD), 완료(COMPLETED)
+
+- [ ] **주문 생성**이 가능하다.
```
주문 생성 시 어떠한 제약 조건들이 있을까요?
조금 더 표현해보면 어떨까요?

hanseu9839 replied:
넵..!

liquidjoo commented on README.md:
```diff
@@ -28,6 +31,92 @@ docker compose -p kitchenpos up -d
+    - [ ] 주문을 생성하면 상태는 **준비(WATTING)**로 설정된다.
+    - [ ] 주문 시간은 모든 검증이 끝난 뒤에 설정한다.
+
+- [ ] **주문 상태**는 변경할 수 있다.
```
주문 상태를 변경할 수 있다 라는 요구사항을 보고 개발을 한다면 상태 값이 언제 변경이 되어야하는지 알기 어려울 것 같아요!
상태 값들이 언제, 어떤 값으로 바뀌는지 조금 더 자세히 나타내보면 좋을 것 같아요

hanseu9839 replied:
조금 더 세분화하여 절차대로 표기하겠습니다..!

liquidjoo commented on README.md:
```diff
@@ -28,6 +31,92 @@ docker compose -p kitchenpos up -d
+            - [ ] **주문 내역의 수량**은 필수이다.
+            - [ ] 사용 가능한 **식사 테이블**이 있어야 한다.
+            - [ ] 위 조건을 만족하면 **식사 테이블**을 할당한다.
+    - [ ] **주문 내역**은 필수이며, 해당 내역의 메뉴들은 **모두 활성화**되어 있어야 한다.
+    - [ ] 주문 내역의 **가격**과 실제 **메뉴의 가격**이 다르면 주문할 수 없다.
```
주문 내역인 OrderLineItem 에 대해서도 조금 더 자세히 표현해 볼 수 있을 것 같군요 :)

hanseu9839 replied:
넵..!

liquidjoo commented on src/main/java/calculate/domain/Number.java:
```diff
@@ -3,30 +3,32 @@
 
     public Number(final String value) {
-        this.value = isValidate(makeNumber(value));
+        this.value = makeNumber(value);
+        validateNonNegativeValue(this.value);
```
```suggestion
        validateNonNegativeValue(this.value);
        this.value = makeNumber(value);
```
유효성을 먼저 검사하는 것도 좋은 방법일 것 같네요 :)

hanseu9839 replied:
넵..!

liquidjoo commented on src/main/java/calculate/domain/Number.java:
```diff
@@ -3,30 +3,32 @@
     }
 
-    private int isValidate(int value) {
+    private void validateNonNegativeValue(int value) {
```
피드백 반영 👍

liquidjoo commented on src/main/java/calculate/domain/NumberGroupsParser.java:
```diff
@@ -0,0 +1,44 @@
+
+public class NumberGroupsParser {
+
+    private NumberGroupsParser() {
+    }
+
+    private static final String CUSTOM_DELIMITER_PREFIX = "//";
+    private static final String DEFAULT_DELIMITER = "[,:]";
+    private static final String CUSTOM_DELIMITER = "//(.)\\n(.*)";
+    private static final Pattern PATTERN = Pattern.compile(CUSTOM_DELIMITER);
```
생성자는 필드 초기화 이후에 하는 것이 좋습니다!

hanseu9839 replied:
넵..!

liquidjoo reviewed (changes requested):
안녕하세요 승균님
미션 구현 잘해주셨어요 👍 
몇 가지 코멘트 남겼습니다~
확인하고 다시 요청해 주세요 🙇

hanseu9839 committed:
refactor: 2차 리뷰 반영 코드 리팩토링
- NumberGroupsParser 클래스를 StringNumberParser로 이름 변경 및 리팩토링
- NumberGroups 클래스에서 StringNumberParser를 사용하도록 의존성 변경
- Number 클래스에서 문자열 입력 생성자를 제거하고, 숫자(int)만 허용하도록 변경

hanseu9839 committed:
refactor: 2차 리뷰 반영 요구사항 리팩토링

hanseu9839 committed:
refactor: [ ] 누락 추가

liquidjoo commented on README.md:
```diff
@@ -34,89 +34,125 @@ docker compose -p kitchenpos up -d
+    - [ ] 모든 메뉴를 조회할 수 있어야 한다.
+    - [ ] 특정 **UUID** 를 기반으로 메뉴를 조회할 수 있어야 한다.
+
+- [ ] 메뉴를 수정한다.
+    - 메뉴 가격을 변경할 수 있어야 한다.
+        - 변경되는 **가격(price)** 은 0 이상의 양수이어야 한다.
+        - 변경된 **가격(price)** 은 메뉴 상품 가격 합보다 클 수 없어야 한다.
+    - 메뉴의 화면 표시 여부를 변경할 수 있어야 한다.
+        - [ ] 메뉴 가격이 메뉴 상품 가격 합보다 큰 경우, 표시 상태로 변경할 수 없어야 한다.
```
[ ] 괄호를 통일해주면 좋을 것 같습니다 :)

liquidjoo commented on README.md:
```diff
@@ -34,89 +34,125 @@ docker compose -p kitchenpos up -d
-
-- [ ] **메뉴 가격 변경** 시. 가격은 **0 이상의 양수**여야 하고, **메뉴 상품 가격 합보다 클 수 없다**.
-- [ ] **메뉴의 화면 표시 여부**는 변경할 수 있으나, **메뉴 가격이 상품 가격 합보다 큰 경우**에는 표시 상태로 바꿀 수 없다.
+- [ ] 메뉴를 생성한다.
+    - [ ] 각 메뉴는 1개의 메뉴 그룹을 반드시 가져야 한다.
+    - [ ] 각 메뉴는 1개 이상의 메뉴 상품을 반드시 포함해야 하며, 해당 메뉴 상품들은 이미 등록된 상품이어야 한다.
+    - [ ] 각 메뉴는 고유한 **UUID** 를 가져야 한다.
+    - [ ] 메뉴의 **가격(price)** 은 필수로 입력되어야 하며, 0 이상의 양수 값이어야 한다.
+    - [ ] 메뉴의 **이름(name)** 은 필수로 입력되어야 하며, 비속어가 포함되어서는 안 된다.
+
+- [ ] 메뉴를 조회한다.
```
피드백 반영 👍

liquidjoo commented on README.md:
```diff
@@ -34,89 +34,125 @@ docker compose -p kitchenpos up -d
+    -	**COMPLETED (완료)** : 모든 처리가 끝난 완료 상태.
+
+- [ ] 주문을 생성한다.
+    - [ ] 주문은 3가지 타입이 존재하며, 타입 필드는 필수이다.
+        - **배달 (DELIVERY)**
+            - [ ] 배송지 주소가 반드시 입력되어야 한다.
+            - [ ] 입력된 배송지 주소가 유효한 경우에만 주문이 생성 가능하다.
+        - **포장 (TAKEOUT)**
+            - [ ] 배송지나 식사 테이블이 필요하지 않다.
+        - **매장 식사 (EAT_IN)**
+            - [ ] 유효한 식사 테이블이 할당되어 있어야 한다.
+            - [ ] 식사 테이블은 반드시 사용 가능한 상태여야 한다.
```
타입인 경우 생성할때만 사용하는 것이 아닌 상태 변경 등에서도 사용이 될 것 같은데요
생성의 하위에 두는 것 보다 같은 레벨에 두어도 좋을 것 같아요 or 타입을 분리 해도 좋습니다

liquidjoo commented on README.md:
```diff
@@ -34,89 +34,125 @@ docker compose -p kitchenpos up -d
+            - [ ] 해당 테이블의 모든 주문이 **주문 완료(COMPLETED)** 상태일 경우, 테이블을 초기화한다.
+            - [ ] 테이블의 손님 수를 0으로 설정하고, **비어있음(occupied = false)** 상태로 변경한다.
+
+### 주문 항목 (OrderLineItem) : 각 주문에 대한 세부 사항
+- 각 주문 항목은 다음 정보를 포함한다:
+    - [ ] 해당 주문 항목에 대한 메뉴 식별자는 필수이다.
+    - [ ] 주문한 메뉴의 개수는 0 이상이어야 한다.
+    - [ ] 주문 항목의 가격은 실제 메뉴 가격과 일치해야 한다.
```
피드백 반영 👍

liquidjoo commented on src/main/java/calculate/domain/StringNumberParser.java:
```diff
@@ -4,41 +4,48 @@
     private static final Pattern PATTERN = Pattern.compile(CUSTOM_DELIMITER);
 
-    public static List<Number> parse(String input) {
+    private StringNumberParser() {
+    }
```
피드백 반영 좋습니다~

liquidjoo reviewed (approved):
안녕하세요 승균님
피드백 반영 잘해주셨어요 👍  몇 가지 코멘트 남겼습니다
확인하고 다음 단계에 반영해보면 좋을 것 같아요 :)
다음 단계 진행해 주세요 🙇
